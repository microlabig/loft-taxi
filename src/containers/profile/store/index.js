export { default } from './reducer';

export * from './actions';
//export * from './middlewares';
export * from './sagas';
export * from './selectors';