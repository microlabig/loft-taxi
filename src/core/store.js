import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { all } from 'redux-saga/effects';

import userReducer, { userSaga } from '../containers/login/store';
import cardReducer, { cardSaga } from '../containers/profile/store';
import addressReducer, { addressSaga } from '../containers/map/store';

// делаем общий срез редьюсеров
export const rootReducer = combineReducers({
    userReducer,
    cardReducer,
    addressReducer
});

// создаем сагу 
const sagaMiddleware = createSagaMiddleware();

// создаем root-сагу для запуска отдельных саг параллельно
function* rootSaga() {
    yield all([
        userSaga(),
        cardSaga(),
        addressSaga()
    ]);
}

// создаем стор
export default createStore(
    rootReducer,
    // объединяем функции справа-налево
    compose(
        // добавляем созданную выше сагу
        applyMiddleware(sagaMiddleware),
        // подключаем расширение Redux DevTools для дебага redux
        window.__REDUX_DEVTOOLS_EXTENSION__
            ? window.__REDUX_DEVTOOLS_EXTENSION__()
            : noop => noop // для тестов
    )
); 

// запускаем root-сагу
sagaMiddleware.run(rootSaga);
