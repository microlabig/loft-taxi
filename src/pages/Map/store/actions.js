import { createAction } from 'redux-actions';
import * as consts from './consts';

// Route
export const fetchRouteLoading = createAction(consts.ROUTE_LOADING);
export const fetchRouteRequest = createAction(consts.ROUTE_REQUEST);
export const fetchRouteSuccess = createAction(consts.ROUTE_SUCCESS);
export const fetchRouteFailure = createAction(consts.ROUTE_FAILURE);

// AddressList
export const fetchAddressLoading = createAction(consts.ADDRESS_LOADING);
export const fetchAddressListRequest = createAction(consts.ADDRESS_LIST_REQUEST);
export const fetchAddressListSuccess = createAction(consts.ADDRESS_LIST_SUCCESS);
export const fetchAddressListFailure = createAction(consts.ADDRESS_LIST_FAILURE);

export const fetchAddressClearError = createAction(consts.ADDRESS_CLEAR_ERROR);
export const fetchAddressReset = createAction(consts.ADDRESS_RESET);