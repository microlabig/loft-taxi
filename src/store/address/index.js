export { default } from './reducer';

export * from './actions';
export * from './middlewares';
export * from './selectors';