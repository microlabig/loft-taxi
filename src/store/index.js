import { createStore, combineReducers, compose, applyMiddleware } from 'redux';

import user, { userMiddleware } from './user';
import card, { cardMiddleware } from './card';

export default createStore(
    combineReducers({
        user,
        card
    }),
    compose(
        applyMiddleware(userMiddleware),
        applyMiddleware(cardMiddleware),
        window.__REDUX_DEVTOOLS_EXTENSION__
            ? window.__REDUX_DEVTOOLS_EXTENSION__
            : noop => noop
    )
); 