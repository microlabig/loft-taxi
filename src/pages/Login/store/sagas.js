import { takeLatest, call, put, fork } from 'redux-saga/effects';
import * as consts from './consts';
import * as actions from './actions';
import * as api from '../../../utils/api';

// сага обработки промежуточных результатов 
function* fetchUserWorker(action) {
    const { type, payload } = action;
    let apiFunc = null;

    // тип экшена откуда произошел диспатч - логин или регистрация
    // для настройки передачи ф-ии api
    switch (type) {
        // логин
        case consts.USER_LOGIN:
            apiFunc = api.userLogin;
            break;
        // регистрация
        case consts.USER_REGISTER:
            apiFunc = api.userRegister;
            break;
        default:
            return;
    }

    // диспатчим экшен установки флага начало загрузки инф-ии
    yield put(actions.fetchUserLoading());

    try {
        // вызываем ф-ию api и ждем когда прийдет ответ от сервера
        const response = yield call(apiFunc, payload);
        // проверяем статус ответа
        if (response.success) {
            // диспатчим экшен успешного запроса
            yield put(actions.fetchUserSuccess(response));
        } else {
            throw new Error(response.error);
        }
    } catch (error) {
        // диспатчим экшен в случае ошибки запроса
        yield put(actions.fetchUserFailure(error.message));
    }
}

// сага возврата последнего вызова fetchUserWorker в стеке запросов логина пользователя
function* loginWatcher() {
    yield takeLatest(consts.USER_LOGIN, fetchUserWorker);
}

// сага возврата последнего вызова fetchUserWorker в стеке запросов регистрации пользователя
function* registerWatcher() {
    yield takeLatest(consts.USER_REGISTER, fetchUserWorker);
}

// общая сага user data запуска неблокирующих запросов
export function* userSaga() {
    yield fork(loginWatcher, null);
    yield fork(registerWatcher, null);
}