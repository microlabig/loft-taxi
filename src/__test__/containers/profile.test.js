import React from 'react';
import { MemoryRouter } from "react-router-dom";

import { mount } from "enzyme";
import { act, render, fireEvent } from "@testing-library/react";

import { Provider } from 'react-redux'
import { createStore } from 'redux';
import { rootReducer } from "../../core/store";

import ProfilePage from '../../pages/profile';

describe('компонент ProfilePage', () => {
    let initStore = null;
    let wrapper = null;

    // подготавливаем DOM-элемент, куда будем рендерить
    beforeEach(() => { // ..Each - выполняется перед каждой ф-ией it
        initStore = createStore(rootReducer);
        wrapper = mount(
            <MemoryRouter>
                <Provider store={initStore}>
                    <ProfilePage />
                </Provider>
            </MemoryRouter>
        );
    });

    // подчищаем после завершения
    afterEach(() => { // ..Each - выполняется перед каждой ф-ией it
        wrapper.unmount();
        initStore = null;
        wrapper = null;
    });

    describe('инициализация', () => {
        it('монтирование компонента', () => {
            expect(wrapper.find(ProfilePage)).toHaveLength(1);
        });

        it('есть хедер, форма и необходимая текстовая информация', () => {
            expect(wrapper.find('Header')).toHaveLength(1);
            expect(wrapper.find('FormPayment')).toHaveLength(1);
            expect(wrapper.find('h2').text()).toEqual("Профиль");
            expect(wrapper.find('p').at(0).text()).toEqual("Способ оплаты");
        });

        it('есть необходимые поля ввода данных карты и кнопка сохранения', () => {
            expect(wrapper.find('input')).toHaveLength(4);
            expect(wrapper.find('button[data-testid="button-save"]')).toHaveLength(1);
        });
    });

    describe('проверка состояния элементов компонента', () => {
        let values = null;

        // подготавливаем DOM-элемент, куда будем рендерить
        beforeEach(() => { // ..Each - выполняется перед каждой ф-ией it
            values = {
                cardNumber: '1231 2312 3122 3423',
                expiryDate: '01/25',
                cardName: 'NAME SURNAME',
                cvc: '123'
            }
        });

        // подчищаем после завершения
        afterEach(() => { // ..Each - выполняется перед каждой ф-ией it
            values = null;
        });

        it('создание снапшота', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('при монтировании компонента кнопка активна', () => {
            expect(wrapper.find('button[data-testid="button-save"]').prop('disabled')).toBeFalsy();
        });

        it('при вводе данных в разные поля ввода кнопка неактивна', async () => {
            // // @testing-library
            // const onChange=jest.fn((e) => {
            //     console.log(e);
            //     e.target.value = '123';
            // });

            // const { getByTestId } = render(
            //     <MemoryRouter>
            //         <Provider store={initStore}>
            //             <ProfilePage />
            //         </Provider>
            //     </MemoryRouter>
            // );

            // const cardNumberInput = getByTestId(/input-cardNumber/i);
            // const expiryDateInput = getByTestId(/input-expiryDate/i);
            // const cardNameInput = getByTestId(/input-cardName/i);
            // const cvcInput = getByTestId(/input-cvc/i);
            // const button = getByTestId(/button-save/i);

            // await act(async () => {
            //     await fireEvent.change(cardNumberInput, { target: { value: values.cardNumber } });
            // });
            // expect(cardNumberInput.getAttribute('value')).toMatch(values.cardNumber);
            // expect(button.disabled).toBeTruthy();

            // await act(async () => {
            //     await fireEvent.change(cardNumberInput, { target: { value: '' } });
            //     await fireEvent.change(expiryDateInput, { target: { value: values.expiryDate } });
            // });
            // expect(button.disabled).toBeTruthy();

            // await act(async () => {
            //     await fireEvent.change(expiryDateInput, { target: { value: '' } });
            //     await fireEvent.change(cardNameInput, { target: { value: values.cardName } });
            // });
            // expect(button.disabled).toBeTruthy();

            // await act(async () => {
            //     await fireEvent.change(cardNameInput, { target: { value: '' } });
            //     await fireEvent.change(cvcInput, { target: { value: values.cvc } });
            // });
            // expect(button.disabled).toBeTruthy();
            // onChange.mockClear();
            // //jest.resetAllMock();

            const cardNumberInput = wrapper.find('[data-testid="input-cardNumber"]');
            const expiryDateInput = wrapper.find('[data-testid="input-expiryDate"]');
            const cardNameInput = wrapper.find('[data-testid="input-cardName"]');
            const cvcInput = wrapper.find('[data-testid="input-cvc"]');
            const button = wrapper.find('button[data-testid="button-save"]').at(0);

            await act(async () => {
                await cardNumberInput.simulate('change', {
                    target: {
                        name: 'cardNumber',
                        value: values.cardNumber
                    }
                });
            });

            expect(button.prop('disabled')).toBeTruthy();

            await act(async () => {
                await cardNumberInput.simulate('change', {
                    target: {
                        name: 'cardNumber',
                        value: ''
                    }
                });
                await expiryDateInput.simulate('change', {
                    target: {
                        name: 'expiryDate',
                        value: values.expiryDate
                    }
                });
            });

            expect(button.prop('disabled')).toBeTruthy();

            await act(async () => {
                await expiryDateInput.simulate('change', {
                    target: {
                        name: 'expiryDate',
                        value: ''
                    }
                });
                await cardNameInput.simulate('change', {
                    target: {
                        name: 'cardName',
                        value: values.cardName
                    }
                });
            });

            expect(button.prop('disabled')).toBeTruthy();

            await act(async () => {
                await cardNameInput.simulate('change', {
                    target: {
                        name: 'cardName',
                        value: ''
                    }
                });
                await cvcInput.simulate('change', {
                    target: {
                        name: 'cvc',
                        value: values.cvc
                    }
                });
            });

            expect(button.prop('disabled')).toBeTruthy();
        });

        it('при вводе валидных данных во все поля ввода кнопка активна', async () => {
            const cardNumberInput = wrapper.find('[data-testid="input-cardNumber"]');
            const expiryDateInput = wrapper.find('[data-testid="input-expiryDate"]');
            const cardNameInput = wrapper.find('[data-testid="input-cardName"]');
            const cvcInput = wrapper.find('[data-testid="input-cvc"]');
            const button = wrapper.find('button[data-testid="button-save"]');

            await act(async () => {
                await cardNumberInput.simulate('change', {
                    target: {
                        name: 'cardNumber',
                        value: values.cardNumber
                    }
                });
                await expiryDateInput.simulate('change', {
                    target: {
                        name: 'expiryDate',
                        value: values.expiryDate
                    }
                });
                await cardNameInput.simulate('change', {
                    target: {
                        name: 'cardName',
                        value: values.cardName
                    }
                });
                await cvcInput.simulate('change', {
                    target: {
                        name: 'cvc',
                        value: values.cvc
                    }
                });
            });

            expect(button.prop('disabled')).toBeFalsy();
        });

    });
});
