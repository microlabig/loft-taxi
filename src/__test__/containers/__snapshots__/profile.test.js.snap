// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`компонент ProfilePage проверка состояния элементов компонента создание снапшота 1`] = `ReactWrapper {}`;
